include/envoy/network/transport_socket.h

```diff
@@ -120,5 +120,25 @@ class TransportSocket {
 
 typedef std::unique_ptr<TransportSocket> TransportSocketPtr;
 
+/**
+ * A factory for creating transport socket. It will be associated to filter chains and clusters.
+ */
+class TransportSocketFactory {
+public:
+  virtual ~TransportSocketFactory() {}
+
+  /**
+   * @return std::string the value to be used as :scheme header in a HTTP request
+   */
+  virtual const std::string& httpScheme() const PURE;
```

Comment:
I haven't fully gone through this change yet, but from a quick skim this seems odd from an interface perspective. Can you explain why we need this here?

Reply:
@lizan I thought about this more, and IMO the best way of handling this to replace this function with `bool implementsTls()` and then just use that to select the scheme much like it was done before. WDYT? I think that's pretty clean. If we don't like "Tls" because it implies something that is too strong, we can do something like `implementsSecureTransport()`. WDYT?

Reply:
`bool implementsSecureTransport()` SGTM. I did this as `std::string&` since http2 spec doesn't strictly limit scheme to "http" or "https". As long as no one needs that flexibility I'm fine with it.

Reply:
Done.